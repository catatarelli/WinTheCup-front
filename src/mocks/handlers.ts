import { rest } from "msw";
import { REACT_APP_API_URL } from "@env";
import type {
  RegisterData,
  UserCredentials,
} from "../redux/features/user/userTypes";
import {
  mockGetPredictionByIdResponse,
  mockgetPredictionsResponse,
} from "./predictionsMocks";

export const handlers = [
  rest.post(`${REACT_APP_API_URL}/user/register`, async (req, res, ctx) => {
    const user = await req.json<RegisterData>();

    if (!user || user.email === "panchito@gmail.com") {
      return res(ctx.status(409));
    }

    return res(ctx.status(201), ctx.json({ user }));
  }),

  rest.post(`${REACT_APP_API_URL}/user/login`, async (req, res, ctx) => {
    const user = await req.json<UserCredentials>();
    if (user.password === "testPassword") {
      return res(ctx.status(401));
    }

    return res(ctx.status(200), ctx.json({ token: "token" }));
  }),

  rest.get(`${REACT_APP_API_URL}/predictions`, async (req, res, ctx) =>
    res.once(
      ctx.status(404),
      ctx.json({ error: "There was an error loading your predictions" })
    )
  ),

  rest.get(`${REACT_APP_API_URL}/predictions`, async (req, res, ctx) =>
    res.once(
      ctx.status(404),
      ctx.json({ error: "There was loading more predictions" })
    )
  ),

  rest.get(`${REACT_APP_API_URL}/predictions`, (req, res, ctx) =>
    res(ctx.status(200), ctx.json(mockgetPredictionsResponse))
  ),

  rest.get(`${REACT_APP_API_URL}/predictions/:predictionId`, (req, res, ctx) =>
    res.once(
      ctx.status(400),
      ctx.json({ error: "There was an error on the server" })
    )
  ),

  rest.get(`${REACT_APP_API_URL}/predictions/:predictionId`, (req, res, ctx) =>
    res(ctx.status(200), ctx.json(mockGetPredictionByIdResponse))
  ),

  rest.post(`${REACT_APP_API_URL}/predictions/create`, (req, res, ctx) =>
    res.once(
      ctx.status(400),
      ctx.json({ error: "There was an error on the server" })
    )
  ),

  rest.post(`${REACT_APP_API_URL}/predictions/create`, (req, res, ctx) =>
    res(ctx.status(201))
  ),

  rest.delete(
    `${REACT_APP_API_URL}/predictions/delete/:predictionId`,
    (req, res, ctx) =>
      res.once(
        ctx.status(404),
        ctx.json({ error: "There was an error on the server" })
      )
  ),

  rest.delete(
    `${REACT_APP_API_URL}/predictions/delete/:predictionId`,
    (req, res, ctx) => res(ctx.status(201))
  ),

  rest.patch(
    `${REACT_APP_API_URL}/predictions/update/:predictionId`,
    (req, res, ctx) =>
      res.once(
        ctx.status(404),
        ctx.json({ error: "There was an error on the server" })
      )
  ),

  rest.patch(
    `${REACT_APP_API_URL}/predictions/update/:predictionId`,
    (req, res, ctx) => res(ctx.status(200))
  ),
];
