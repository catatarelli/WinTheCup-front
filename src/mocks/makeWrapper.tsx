import React from "react";
import { Provider } from "react-redux";
import { store } from "../redux/store";

const makeWrapper = ({ children }: { children: JSX.Element }) => (
  <Provider store={store}>{children}</Provider>
);

export default makeWrapper;
