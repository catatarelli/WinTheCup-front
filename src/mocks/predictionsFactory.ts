/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { Factory } from "fishery";
import { faker } from "@faker-js/faker";
import type { PredictionStructure } from "../redux/features/user/userTypes";

const predictionFactory = Factory.define<PredictionStructure>(() => ({
  match: "Argentina vs England",
  goalsTeam1: faker.datatype.number(),
  goalsTeam2: faker.datatype.number(),
  redCards: faker.datatype.number(),
  yellowCards: faker.datatype.number(),
  penalties: faker.datatype.number(),
  picture: faker.image.sports(),
  backupPicure: faker.image.sports(),
}));

export const getRandomPredictionsList = (number: number) =>
  predictionFactory.buildList(number);
