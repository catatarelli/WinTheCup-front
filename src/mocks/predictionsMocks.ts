import { type ImageSourcePropType } from "react-native";
import {
  type PredictionsState,
  type PredictionsResponse,
  type PredictionStructure,
} from "../redux/features/predictions/predictionsTypes";
import {
  getRandomPrediction,
  getRandomPredictionsList,
} from "./predictionsFactory";

export const mockgetPredictionsResponse: PredictionsResponse = {
  predictions: getRandomPredictionsList(10) as PredictionStructure[],
};

export const emptyPredictionsState = {
  predictions: [],
  currentPrediction: {
    match: "",
    goalsTeam1: 0,
    goalsTeam2: 0,
    redCards: 0,
    yellowCards: 0,
    penalties: 0,
    picture: "" as ImageSourcePropType,
    createdBy: "",
    id: "",
  },
};

export const currentPredictionMock: PredictionStructure = {
  match: "Argentina vs Mexico",
  goalsTeam1: 3,
  goalsTeam2: 2,
  createdBy: "543",
  id: "234",
  picture: "asdf" as ImageSourcePropType,
};

export const mockPredictionsState: PredictionsState = {
  predictions: getRandomPredictionsList(10) as PredictionStructure[],
  currentPrediction: currentPredictionMock,
};

export const mockGetPredictionByIdResponse: PredictionStructure = {
  ...(getRandomPrediction() as PredictionStructure),
  id: "56789",
};
