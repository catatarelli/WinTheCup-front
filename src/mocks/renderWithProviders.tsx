import React, { type PropsWithChildren } from "react";
import { render } from "@testing-library/react-native";
import type { RenderOptions } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { type store, type RootState } from "../redux/store";
import { uiReducer } from "../redux/features/ui/uiSlice";
import { userReducer } from "../redux/features/user/userSlice";
import { NavigationContainer } from "@react-navigation/native";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: typeof store;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState,
    store = configureStore({
      reducer: { ui: uiReducer, user: userReducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = ({
    children,
  }: PropsWithChildren<Record<string, unknown>>): JSX.Element => (
    <>
      <Provider store={store}>
        <NavigationContainer>{children}</NavigationContainer>
      </Provider>
    </>
  );

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
