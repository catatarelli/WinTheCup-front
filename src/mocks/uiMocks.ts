import type { UiState } from "../redux/features/uiSlice/uiTypes.js";

export const emptyModalMock: UiState = {
  modal: "",
  isError: false,
};

export const mockUiStateSuccessModal: UiState = {
  modal: "Account created successfully",
  isError: false,
};

export const mockUiStateErrorModal: UiState = {
  modal: "User is already registered",
  isError: true,
};
