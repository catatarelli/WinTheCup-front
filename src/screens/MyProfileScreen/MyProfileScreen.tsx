import { SafeAreaView, ScrollView, Text } from "react-native";
import React from "react";
import screenStyles from "../../styles/screen.styles";

import headingStyles from "../../styles/headings.styles";
import MyProfile from "../../components/MyProfile/MyProfile";

const MyProfileScreen = (): JSX.Element => (
  <ScrollView style={screenStyles.screenBackground}>
    <SafeAreaView style={screenStyles.container}>
      <Text style={headingStyles.pageTitle}>My Profile</Text>
      <MyProfile />
    </SafeAreaView>
  </ScrollView>
);

export default MyProfileScreen;
