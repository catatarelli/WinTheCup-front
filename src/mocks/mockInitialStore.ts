import { configureStore } from "@reduxjs/toolkit";
import { uiReducer } from "../redux/features/ui/uiSlice";
import { userReducer } from "../redux/features/user/userSlice";
import { store } from "../redux/store";
import { mockUiStateErrorModal, mockUiStateSuccessModal } from "./uiMocks";
import { loggedInUserStateMock } from "./userMocks";

export const mockInitialStoreSuccessModal: typeof store = configureStore({
  reducer: {
    ui: uiReducer,
    user: userReducer,
  },
  preloadedState: {
    ui: mockUiStateSuccessModal,
    user: loggedInUserStateMock,
  },
});

export const mockInitialStoreErrorModal: typeof store = configureStore({
  reducer: {
    ui: uiReducer,
    user: userReducer,
  },
  preloadedState: {
    ui: mockUiStateErrorModal,
  },
});
