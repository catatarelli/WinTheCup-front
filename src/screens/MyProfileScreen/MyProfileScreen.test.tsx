import React from "react";
import { screen, render } from "@testing-library/react-native";
import { Provider } from "react-redux";
import { mockInitialStoreSuccessModal } from "../../mocks/mockInitialStore";
import { NavigationContainer } from "@react-navigation/native";
import MyProfileScreen from "./MyProfileScreen";

beforeEach(() => {
  jest.clearAllMocks();
});

jest.mock("@react-native-async-storage/async-storage", () => ({
  setItem: jest.fn(),
  getItem: jest.fn(),
  removeItem: jest.fn(),
}));

describe("Given a MyProfileScreen component", () => {
  describe("When it's rendered", () => {
    test("Then it should show the title 'My Profile'", async () => {
      const titleText = "My Profile";

      render(
        <Provider store={mockInitialStoreSuccessModal}>
          <NavigationContainer>
            <MyProfileScreen />
          </NavigationContainer>
        </Provider>
      );

      const title = screen.queryByText(titleText);

      expect(title).toBeDefined();
    });

    test("Then it should match the snapshot", () => {
      const screen = render(
        <Provider store={mockInitialStoreSuccessModal}>
          <NavigationContainer>
            <MyProfileScreen />
          </NavigationContainer>
        </Provider>
      );
      expect(screen).toMatchSnapshot();
    });
  });
});
