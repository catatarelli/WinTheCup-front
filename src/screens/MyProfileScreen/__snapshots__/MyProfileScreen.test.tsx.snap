// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a MyProfileScreen component When it's rendered Then it should match the snapshot 1`] = `
<RCTScrollView
  style={
    Object {
      "backgroundColor": "#fff",
      "height": "100%",
    }
  }
>
  <View>
    <RCTSafeAreaView
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#3F3F3F",
            "fontSize": 30,
            "marginBottom": 10,
            "marginTop": 50,
          }
        }
      >
        My Profile
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#EEEEE4",
            "borderRadius": 15,
            "display": "flex",
            "elevation": 6,
            "height": 612,
            "margin": 10,
            "marginBottom": 50,
            "marginVertical": 15,
            "paddingVertical": 15,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 3,
              "width": 0,
            },
            "shadowOpacity": 0.27,
            "shadowRadius": 4.65,
            "width": 321,
          }
        }
      >
        <View>
          <Text
            style={
              Object {
                "color": "#3F3F3F",
                "fontSize": 20,
              }
            }
          >
            Username
          </Text>
          <TextInput
            maxLength={20}
            onChangeText={[Function]}
            style={
              Object {
                "borderColor": "#3F3F3F",
                "borderStyle": "solid",
                "borderWidth": 2,
                "color": "#3F3F3F",
                "fontSize": 20,
                "height": 54,
                "marginVertical": 10,
                "padding": 10,
                "width": 300,
              }
            }
            testID="username"
            textContentType="username"
            value="luis123"
          />
          <Text
            style={
              Object {
                "color": "#3F3F3F",
                "fontSize": 20,
              }
            }
          >
            Password
          </Text>
          <TextInput
            maxLength={20}
            onChangeText={[Function]}
            secureTextEntry={true}
            style={
              Object {
                "borderColor": "#3F3F3F",
                "borderStyle": "solid",
                "borderWidth": 2,
                "color": "#3F3F3F",
                "fontSize": 20,
                "height": 54,
                "marginVertical": 10,
                "padding": 10,
                "width": 300,
              }
            }
            testID="password"
            textContentType="password"
            value=""
          />
          <Text
            style={
              Object {
                "color": "#3F3F3F",
                "fontSize": 20,
              }
            }
          >
            Email
          </Text>
          <TextInput
            maxLength={25}
            onChangeText={[Function]}
            style={
              Object {
                "borderColor": "#3F3F3F",
                "borderStyle": "solid",
                "borderWidth": 2,
                "color": "#3F3F3F",
                "fontSize": 20,
                "height": 54,
                "marginVertical": 10,
                "padding": 10,
                "width": 300,
              }
            }
            testID="email"
            textContentType="emailAddress"
            value="luis@gmail.com"
          />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#3F3F3F",
              "borderRadius": 20,
              "color": "#fff",
              "height": 57,
              "justifyContent": "center",
              "marginBottom": 10,
              "marginTop": 10,
              "marginVertical": 25,
              "opacity": 0.8,
              "padding": 6,
              "width": 160,
              "zIndex": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#fff",
                "fontSize": 20,
              }
            }
            testID="saveChanges"
          >
            Save changes
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#EEEEE4",
              "borderColor": "#8A1538",
              "borderRadius": 20,
              "borderWidth": 3,
              "color": "#fff",
              "height": 57,
              "justifyContent": "center",
              "marginBottom": 10,
              "marginTop": 120,
              "marginVertical": 25,
              "opacity": 0.8,
              "padding": 6,
              "width": 160,
              "zIndex": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#8A1538",
                "fontSize": 20,
              }
            }
          >
            Log out
          </Text>
        </View>
      </View>
    </RCTSafeAreaView>
  </View>
</RCTScrollView>
`;
